runtime/shibuya/src/lib.rs

```diff
@@ -2238,53 +2241,45 @@ impl_runtime_apis! {
         }
     }
 
-    impl xcm_fee_payment_runtime_api::XcmPaymentApi<Block> for Runtime {
+    impl xcm_fee_payment_runtime_api::fees::XcmPaymentApi<Block> for Runtime {
         fn query_acceptable_payment_assets(xcm_version: xcm::Version) -> Result<Vec<VersionedAssetId>, XcmPaymentApiError> {
             if !matches!(xcm_version, xcm::v3::VERSION | xcm::v4::VERSION) {
                 return Err(XcmPaymentApiError::UnhandledXcmVersion);
             }
 
             // Native asset is always supported
-            let native_asset_location: XcmLocation = XcmLocation::try_from(xcm_config::ShibuyaLocation::get())
-            .map_err(|_| XcmPaymentApiError::VersionedConversionFailed)?;
-
-            Ok([VersionedAssetId::V4(native_asset_location.into())]
-                .into_iter()
-                // Acquire foreign assets which have 'units per second' configured
-                .chain(
-                    pallet_xc_asset_config::AssetLocationUnitsPerSecond::<Runtime>::iter_keys().filter_map(|asset_location| {
-
-                        match XcmLocation::try_from(asset_location) {
-                            Ok(asset) => Some(VersionedAssetId::V4(asset.into())),
-                            Err(_) => None,
-                        }
-                    })
-            ).filter_map(|asset| asset.into_version(xcm_version).ok()).collect())
+            let mut acceptable_assets = vec![XcmAssetId::from(xcm_config::ShibuyaLocation::get())];
+
+            // Add foreign assets that have 'units per second' configured
+            acceptable_assets.extend(
+                pallet_xc_asset_config::AssetLocationUnitsPerSecond::<Runtime>::iter_keys().filter_map(
+                    |asset_location| match XcmLocation::try_from(asset_location) {
+                        Ok(location) => Some(XcmAssetId::from(location)),
+                        Err(_) => None,
+                    },
+                ),
+            );
+
+            PolkadotXcm::query_acceptable_payment_assets(xcm_version, acceptable_assets)
         }
 
-        // TODO: improve this function, reduce code duplication, especially on a such a functional level
```

Comment:
I see you tried to resolve this with a good step forward but it's not what I had in mind 🙂 

Right now the `WeightToForeignAssetFee` is in no way _bound_ to the pallet that actually handles the conversion, which is `XcmAssetConfig`.

If you want to take this approach, I'd suggest to define this function directly in that pallet. That way it would be more _tied_ together. Because even though you've introduced the new type with the function, it's not used inside aforementioned pallet, hence we still have unnecessary code (and functional) duplication.

Reply:
I have moved the `weight_to_fee` logic directly into the `XcAssetConfig` pallet. In the future, we could consider creating a `WeightToFee` trait (similar to `ExecutionPaymentRate`) to eliminate code duplication in *astar-primitives*. This might be outside the scope of this uplift, but it's something to keep in mind.